File: CS203 Data Structures and Algorithm Analysis/Template/Misc/FFT.java
import java.util.Scanner;

public class Main {
    public static int[] readArray(Scanner scanner) {
        int n = scanner.nextInt();
        int[] x = new int[n];
        for (int i = n - 1; i >= 0; i--)
            x[i] = scanner.nextInt();
        return x;
    }

    public static void fftTransform(Complex[] a, int n, int k, Complex[] omega) {
        for (int i = 0; i < n; i++) {
            int t = 0;
            for (int j = 0; j < k; j++)
                if ((i & (1 << j)) > 0)
                    t |= (1 << (k - j - 1));
            if (i < t) {
                Complex tmp = a[i];
                a[i] = a[t];
                a[t] = tmp;
            }
        }

        for (int j = 2; j <= n; j *= 2) {
            int m = j / 2;
            for (int p = 0; p < n; p += j) {
                for (int i = 0; i < m; i++) {
                    Complex t = omega[n / j * i].multiply(a[p + m + i]);
                    a[p + m + i] = a[p + i].minus(t);
                    a[p + i] = a[p + i].add(t);
                }
            }
        }
    }

    public static int[] multiply(int[] x, int[] y) {
        int k = 0;
        while ((1 << k) < x.length + y.length)
            k++;
        int n = 1 << k;

        Complex[] omega = new Complex[n];
        Complex[] omegaInv = new Complex[n];
        for (int i = 0; i < n; i++) {
            omega[i] = Complex.omega(n, i);
            omegaInv[i] = omega[i].conj();
        }

        Complex[] c1 = new Complex[n];
        Complex[] c2 = new Complex[n];
        for (int i = 0; i < n; i++)
            c1[i] = i < x.length ? new Complex(x[i], .0) : Complex.ZERO;
        for (int i = 0; i < n; i++)
            c2[i] = i < y.length ? new Complex(y[i], .0) : Complex.ZERO;

        fftTransform(c1, n, k, omega);
        fftTransform(c2, n, k, omega);

        for (int i = 0; i < n; i++)
            c1[i] = c1[i].multiply(c2[i]);

        fftTransform(c1, n, k, omegaInv);

        int[] z = new int[ x.length + y.length - 1 ];
        for (int i = 0; i < z.length; i++)
            z[i] = (int)Math.floor(c1[i].real / n + 0.5);
        return z;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        while (t-- > 0) {
            int[] x = readArray(scanner);
            int[] y = readArray(scanner);

            int[] z = multiply(x, y);
            for (int i = z.length - 1; i >= 0; i--)
                System.out.print(z[i] + " ");
            System.out.println();
        }
    }
}

class Complex {
    public static final Complex ZERO = new Complex(0, 0);

    double real, imag;

    public Complex(double real, double imag) {
        this.real = real;
        this.imag = imag;
    }

    public static Complex omega(int n, int i) {
        return new Complex(Math.cos(2 * Math.PI / n * i), Math.sin(2 * Math.PI / n * i));
    }

    public Complex conj() {
        return new Complex(this.real, -this.imag);
    }

    public Complex add(Complex other) {
        return new Complex(this.real + other.real, this.imag + other.imag);
    }

    public Complex minus(Complex other) {
        return new Complex(this.real - other.real, this.imag - other.imag);
    }

    public Complex multiply(Complex other) {
        return new Complex(
                this.real * other.real - this.imag * other.imag,
                this.real * other.imag + this.imag * other.real
        );
    }
}

File: CS203 Data Structures and Algorithm Analysis/Template/Misc/FFT.cpp
#include <iostream>
#include <cmath>
using namespace std;

const int MAXN = 1e5 + 5;
long long x[MAXN], y[MAXN], z[MAXN], n, m;

class Complex {
    public:
    double real, imag;

    Complex() : real(0), imag(0) {}
    Complex(double r, double i) : real(r), imag(i) {}
    Complex operator + (const Complex& c) {
        return Complex(real + c.real, imag + c.imag);
    }
    Complex operator - (const Complex& c) {
        return Complex(real - c.real, imag - c.imag);
    }
    Complex operator * (const Complex& c) {
        return Complex(
            real * c.real - imag * c.imag,
            real * c.imag + imag * c.real
        );
    }
    Complex conj() {
        return Complex(real, -imag);
    }
};

const Complex ZERO = Complex();
Complex Omega(int n, int k) {
    return Complex(cos(2 * M_PI * k / n), sin(2 * M_PI * k / n));
}

void fft(Complex* a, int n, int k, Complex* omega) {
    for(int i = 0; i < n; i++) {
        int t = 0;
        for(int j = 0; j < k; j++) if((i & (1 << j)) != 0) t |= 1 << (k - j - 1);
        if(i < t) {
            Complex tmp = a[i];
            a[i] = a[t];
            a[t] = tmp;
        }
    }
    for(int j = 2; j <= n; j *= 2) {
        int m = j >> 1;
        for(int p = 0; p < n; p += j) {
            for(int i = 0; i < m; i++) {
                Complex t = omega[n / j * i] * a[p + m + i];
                a[p + m + i] = a[p + i] - t;
                a[p + i] = a[p + i] + t;
            }
        }
    }
    
}

void multiply() {
    int k = 0;
    while((1 << k) < n + m) k++;
    int n = 1 << k;
    Complex* omega = new Complex[n];
    Complex* omegaInv = new Complex[n];
    Complex* a = new Complex[n];
    Complex* b = new Complex[n];
    for(int i = 0; i < n; i++) {
        omega[i] = Omega(n, i);
        omegaInv[i] = omega[i].conj();
        a[i] = i < n ? Complex(x[i], 0) : ZERO;
        b[i] = i < m ? Complex(y[i], 0) : ZERO;
    }
    fft(a, n, k, omega);
    fft(b, n, k, omega);
    for(int i = 0; i < n; i++) a[i] = a[i] * b[i];
    fft(a, n, k, omegaInv);
    for(int i = 0; i < n; i++) z[i] = (long long) floor(a[i].real / n + 0.5);
}

int main() {
    ios::sync_with_stdio(false), cin.tie(0), cout.tie(0);
    cin >> n >> m;
    for(int i = n - 1; i >= 0; i--) cin >> x[i];
    for(int i = m - 1; i >= 0; i--) cin >> y[i];
    multiply();
    for(int i = n + m - 2; i >= 0; i--) cout << z[i] << (i == 0 ? "\n" : " ");

    return 0;
}